ygot/diff_test.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ygot
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/kylelemons/godebug/pretty"
+	gnmipb "github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/openconfig/ygot/util"
+)
+
+func TestSchemaPathToGNMIPath(t *testing.T) {
+	tests := []struct {
+		desc string
+		in   []string
+		want *gnmipb.Path
+	}{{
+		desc: "single element",
+		in:   []string{"one"},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}},
+		},
+	}, {
+		desc: "multiple elements",
+		in:   []string{"one", "two", "three"},
+		want: &gnmipb.Path{
+			Elem: []*gnmipb.PathElem{{
+				Name: "one",
+			}, {
+				Name: "two",
+			}, {
+				Name: "three",
+			}},
+		},
+	}}
+
+	for _, tt := range tests {
+		if got := schemaPathTogNMIPath(tt.in); !proto.Equal(got, tt.want) {
```

Comment:
(optional) Do we usually test unexported functions? Would it be better from maintenance perspective to test all the behaviors via the public API? If an unexported function is complex enough to need its own test, maybe we should consider placing it in its own library.

Reply:
Throughout this library we do a lot of testing of unexported functions. The issue is that we end up with quite a lot of complexity that hides behind very few public APIs -- so we'd rather have some tests here that allow more easy isolation of issues for debugging. 

The logic we applied as to whether something is its own library is whether we really expect to call it from somewhere else -- `util` is generally a collection of things that fall into that category.

I'd prefer to keep the testing like this if we can -- it's a tradeoff between some duplication of testing, but more precise error localisation when we end up with regression issues IMHO.